File: 백준/Gold/2250. 트리의 높이와 너비/트리의 높이와 너비.cc
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
#define ordered_set tree<int, null_type, less<int>, rb_tree_tag,tree_order_statistics_node_update>
#define fastio cin.tie(0), cout.tie(0), ios::sync_with_stdio(0);
#define all(x) (x).begin(), (x).end()
#define x first 
#define y second
using namespace std; typedef long long ll;
using ld = long double;
using ull = unsigned long long;
using pll = pair<ll, ll>; using tll = tuple<ll, ll, ll>;
ll n, m, k, t; string s;

constexpr ll INF = 0x3f3f3f3f3f3f3f3f;
constexpr ll MAX = 10101;
constexpr ll MOD = 998244353;
ll l[MAX], r[MAX];
ll mn[MAX], mx[MAX], seq;
bool root[MAX];

void dfs(ll cur, ll dep){
    if(l[cur] != -1) dfs(l[cur], dep + 1);
    ++seq;
    mn[dep] = min(mn[dep], seq);
    mx[dep] = max(mx[dep], seq);
    if(r[cur] != -1) dfs(r[cur], dep + 1);
}


int main() {
    fastio;

    cin >> n;
    memset(root, 1, sizeof(root));
    memset(mn, 0x3f, sizeof(mn));

    for(int i = 1;i <= n;i++){
        ll idx; cin >> idx;
        cin >> l[idx] >> r[idx];

        if(l[idx] != -1) root[l[idx]] = 0;
        if(r[idx] != -1) root[r[idx]] = 0;
    }

    ll st = 0;
    for(int i = 1;i <= n;i++) if(root[i]) st = i;
    dfs(st, 1);

    ll result = 0, idx = 0;
    for(int i = 1;i <= n;i++){
        if(mn[i] == INF) break;
        if(result >= mx[i] - mn[i] + 1) continue;
        result = mx[i] - mn[i] + 1; idx = i;
    }

    cout << idx << " " << result;


    return 0;
}

File: 백준/Gold/2250. 트리의 높이와 너비/트리의 높이와 너비.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

inf = 1000000000
n = int(input().rstrip())
l = [0] * (n + 1)
r = [0] * (n + 1)
idx = [0] * (n + 1)
root = [1] * (n + 1)
mx = [0] * (n + 1)
mn = [inf] * (n + 1)
num = 0

for _ in range(n):
    arr = list(map(int, input().rstrip().split()))
    idx = arr[0]
    l[idx], r[idx] = arr[1], arr[2]
    
    if l[idx] != -1:
       root[l[idx]] = 0
    if r[idx] != -1:
        root[r[idx]] = 0

def dfs(cur, dep):
    global num
    if l[cur] != -1:
        dfs(l[cur], dep + 1)

    num += 1
    mn[dep] = min(mn[dep], num)
    mx[dep] = max(mx[dep], num)

    if r[cur] != -1:
        dfs(r[cur], dep + 1)

for i in range(1, n + 1):
    if root[i] == 0:
        continue
    dfs(i, 1)

result = 0
diff = 0

for i in range(1, n + 1):
    if mx[i] - mn[i] + 1 > diff:
        diff = mx[i] - mn[i] + 1
        result = i

print(f"{result} {diff}")